source/validate_types.cpp

```diff
@@ -301,85 +304,29 @@ bool ValidationState_t::in_block() const {
-    case SpvCapabilitySampledRect:                       out = SpvCapabilityShader;           break;
-    case SpvCapabilityGenericPointer:                    out = SpvCapabilityAddresses;        break;
-    case SpvCapabilityInt8:                              out = SpvCapabilityKernel;           break;
-    case SpvCapabilityInputAttachment:                   out = SpvCapabilityShader;           break;
-    case SpvCapabilitySparseResidency:                   out = SpvCapabilityShader;           break;
-    case SpvCapabilityMinLod:                            out = SpvCapabilityShader;           break;
-    case SpvCapabilitySampled1D:                         out = SpvCapabilityShader;           break;
-    case SpvCapabilityImage1D:                           out = SpvCapabilitySampled1D;        break;
-    case SpvCapabilitySampledCubeArray:                  out = SpvCapabilityShader;           break;
-    case SpvCapabilitySampledBuffer:                     out = SpvCapabilityShader;           break;
-    case SpvCapabilityImageBuffer:                       out = SpvCapabilitySampledBuffer;    break;
-    case SpvCapabilityImageMSArray:                      out = SpvCapabilityShader;           break;
-    case SpvCapabilityStorageImageExtendedFormats:       out = SpvCapabilityShader;           break;
-    case SpvCapabilityImageQuery:                        out = SpvCapabilityShader;           break;
-    case SpvCapabilityDerivativeControl:                 out = SpvCapabilityShader;           break;
-    case SpvCapabilityInterpolationFunction:             out = SpvCapabilityShader;           break;
-    case SpvCapabilityTransformFeedback:                 out = SpvCapabilityShader;           break;
-    case SpvCapabilityGeometryStreams:                   out = SpvCapabilityGeometry;         break;
-    case SpvCapabilityStorageImageReadWithoutFormat:     out = SpvCapabilityShader;           break;
-    case SpvCapabilityStorageImageWriteWithoutFormat:    out = SpvCapabilityShader;           break;
-    default:
-      out = cap;
-      break;
-  }
-  return out;
-}
-// clang-format on
-
 void ValidationState_t::registerCapability(SpvCapability cap) {
-  SpvCapability capability = cap;
-  // Set dependant capabilities
-  do {
-    module_capabilities_[capability] = true;
-    capability = getDependantCapability(capability);
-  } while (getDependantCapability(capability) != capability);
-
-  // Set Base capability
-  module_capabilities_[capability] = true;
+  module_capabilities_[cap] = true;
+  spv_operand_desc desc;
+  if (SPV_SUCCESS ==
+      grammar_.lookupOperand(SPV_OPERAND_TYPE_CAPABILITY, cap, &desc))
+    libspirv::ForEach(desc->capabilities,
+                      [this](SpvCapability c) { registerCapability(c); });
 }
 
-bool ValidationState_t::hasCapability(SpvCapability cap) {
+bool ValidationState_t::hasCapability(SpvCapability cap) const {
   return module_capabilities_[cap];
```

Comment:
This begs the question: why not store module_capabilities_ as a bitmask in the first place.  Then HasAnyOf is just a bitwise-AND followed by test against zero.

I'm ok either way.